src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.cs

```diff
@@ -0,0 +1,288 @@
+            var offsetInSnippet = new TextSpan(positionInFullText - snippetPositionInFullText, length);
+            return offsetInSnippet;
+        }
+    }
+
+    /// <summary>
+    /// Create the snippet with the full default text and functions applied.  Output the spans associated with
+    /// each field and the final caret location in that text so that we can find those locations later.
+    /// </summary>
+    private static async Task<(string ReplacedSnippetText, ImmutableDictionary<SnippetFieldPart, ImmutableArray<TextSpan>> Fields, TextSpan? CaretSpan)> GetReplacedSnippetTextAsync(
+        Document originalDocument,
+        SourceText originalSourceText,
+        TextSpan snippetSpan,
+        ParsedXmlSnippet parsedSnippet,
+        CancellationToken cancellationToken)
+    {
+        var documentWithDefaultSnippet = originalDocument.WithText(
+            originalSourceText.WithChanges(new TextChange(snippetSpan, parsedSnippet.DefaultText)));
+
+        // Iterate the snippet parts so that we can do two things:
+        //   1.  Calculate the snippet function result.  This must be done against the document containing the default snippet text
+        //       as the function result is context dependent.
+        //   2.  After inserting the function result, determine the spans associated with each editable snippet field.
+        var fieldOffsets = new Dictionary<SnippetFieldPart, ImmutableArray<TextSpan>>();
+        using var _ = PooledStringBuilder.GetInstance(out var functionSnippetBuilder);
+        TextSpan? caretSpan = null;
+
+        // This represents the field start location in the context of the snippet without functions replaced (see below).
+        var locationInDefaultSnippet = snippetSpan.Start;
+
+        // This represents the field start location in the context of the snippet with functions replaced.
+        var locationInFinalSnippet = snippetSpan.Start;
+        foreach (var originalPart in parsedSnippet.Parts)
+        {
+            var part = originalPart;
+
+            // Adjust the text associated with this snippet part by applying the result of the snippet function.
+            if (part is SnippetFunctionPart functionPart)
+            {
+                // To avoid a bunch of document changes and re-parsing, we always calculate the snippet function result
+                // against the document with the default snippet text applied to it instead of with each incremental function result.
+                // So we need to remember the index into the original document.
+                part = await functionPart.WithSnippetFunctionResultAsync(documentWithDefaultSnippet, new TextSpan(locationInDefaultSnippet, part.DefaultText.Length), cancellationToken).ConfigureAwait(false);
+            }
+
+            // Only store spans for editable fields or the cursor location, we don't need to get back to anything else.
+            if (part is SnippetFieldPart fieldPart && fieldPart.EditIndex != null)
```

Comment:
Can the above function part translate the part into a `SnippetFieldPart` or should this be an `else if`

Reply:
a function part is a `SnippetFieldPart`

this is not an else if since they're doing two different actions
A function is a field, but to get the proper value to insert we need to apply the function based on the snippet default text (what happens above).
Here, we store spans for editable fields.  Note that a function field could also be an editable field so this needs to be its own if.